web3.eth.gas.pas

```diff
@@ -99,36 +192,74 @@ procedure estimateGas(
   client    : IWeb3;
   from, &to : TAddress;
   const data: string;
-  default   : TWei;
+  default   : BigInteger;
   callback  : TAsyncQuantity);
-var
-  obj: TJsonObject;
 begin
-  // construct the transaction call object
-  obj := web3.json.unmarshal(Format(
-    '{"from": %s, "to": %s, "data": %s}', [
-      web3.json.quoteString(string(from), '"'),
-      web3.json.quoteString(string(&to), '"'),
-      web3.json.quoteString(data, '"')
-    ]
-  )) as TJsonObject;
-  try
-    // estimate how much gas is necessary for the transaction to complete (without creating a transaction on the blockchain)
-    client.Call('eth_estimateGas', [obj], procedure(resp: TJsonObject; err: IError)
+  // estimate how much gas is necessary for the transaction to complete (without creating a transaction on the blockchain)
+  var eth_estimateGas := procedure(client: IWeb3; const json: string; default: BigInteger; callback: TAsyncQuantity)
+  begin
+    var obj := web3.json.unmarshal(json) as TJsonObject;
+    try
+      client.Call('eth_estimateGas', [obj], procedure(resp: TJsonObject; err: IError)
+      begin
+        if Assigned(err) then
+        begin
+          if err.Message.Contains('gas required exceeds allowance') and (default > 0) then
+            callback(default, nil)
+          else
+            callback(0, err);
+          EXIT;
+        end;
+        callback(web3.json.getPropAsStr(resp, 'result'), nil);
+      end);
+    finally
+      obj.Free;
+    end;
+  end;
+
+  // construct the eip-1559 transaction call object
+  if client.Chain.TxType >= 2 then
+  begin
+    getMaxPriorityFeePerGas(client, procedure(tip: TWei; err: IError)
     begin
       if Assigned(err) then
-      begin
-        if err.Message.Contains('gas required exceeds allowance') and (default > 0) then
-          callback(default, nil)
-        else
-          callback(0, err);
-        EXIT;
-      end;
-      callback(web3.json.getPropAsStr(resp, 'result'), nil);
+        callback(0, err)
+      else
+        getMaxFeePerGas(client, procedure(maxFee: TWei; err: IError)
+        begin
+          if Assigned(err) then
+            callback(0, err)
+          else
+            eth_estimateGas(client, Format(
+              '{"from": %s, "to": %s, "data": %s, "maxPriorityFeePerGas": %s, "maxFeePerGas": %s}', [
+                web3.json.quoteString(string(from), '"'),
+                web3.json.quoteString(string(&to), '"'),
+                web3.json.quoteString(data, '"'),
+                web3.json.quoteString(toHex(tip, [zeroAs0x0]), '"'),
+                web3.json.quoteString(toHex(maxFee, [zeroAs0x0]), '"')
```

Comment:
I believe you're missing `accessList`

Reply:
@GregTheGreek does `eth_estimateGas` need `accessList`? I assumed this to be associated with type 2 transaction signing only.

Reply:
my mistake!